File: PhotonCore/src/main/java/com/outoftheboxrobotics/photoncore/HAL/Motors/PhotonDcMotor.java
package com.outoftheboxrobotics.photoncore.HAL.Motors;

import com.outoftheboxrobotics.photoncore.HAL.HAL;
import com.outoftheboxrobotics.photoncore.HAL.Motors.Commands.PhotonLynxGetADCCommand;
import com.outoftheboxrobotics.photoncore.HAL.Motors.Commands.PhotonLynxGetMotorChannelCurrentAlertLevelCommand;
import com.outoftheboxrobotics.photoncore.HAL.Motors.Commands.PhotonLynxGetMotorChannelEnableCommand;
import com.outoftheboxrobotics.photoncore.HAL.Motors.Commands.PhotonLynxGetMotorChannelModeCommand;
import com.outoftheboxrobotics.photoncore.HAL.Motors.Commands.PhotonLynxGetMotorPIDControlLoopCoefficientsCommand;
import com.outoftheboxrobotics.photoncore.HAL.Motors.Commands.PhotonLynxGetMtorPIDFControlLoopCoefficientsCommand;
import com.outoftheboxrobotics.photoncore.HAL.Motors.Commands.PhotonLynxSetMotorChannelCurrentAlertLevelCommand;
import com.outoftheboxrobotics.photoncore.HAL.Motors.Commands.PhotonLynxSetMotorChannelEnableCommand;
import com.outoftheboxrobotics.photoncore.HAL.Motors.Commands.PhotonLynxSetMotorChannelModeCommand;
import com.outoftheboxrobotics.photoncore.HAL.Motors.Commands.PhotonLynxSetMotorConstantPowerCommand;
import com.outoftheboxrobotics.photoncore.HAL.Motors.Commands.PhotonLynxSetMotorPIDFControlLoopCoefficientsCommand;
import com.outoftheboxrobotics.photoncore.HAL.Motors.Commands.PhotonLynxSetMotorTargetPositionCommand;
import com.outoftheboxrobotics.photoncore.HAL.Motors.Commands.PhotonLynxSetMotorTargetVelocityCommand;
import com.outoftheboxrobotics.photoncore.HAL.PhotonHAL;
import com.outoftheboxrobotics.photoncore.hardware.PhotonLynxDcMotorController;
import com.qualcomm.hardware.lynx.LynxNackException;
import com.qualcomm.hardware.lynx.commands.LynxCommand;
import com.qualcomm.hardware.lynx.commands.LynxMessage;
import com.qualcomm.hardware.lynx.commands.core.LynxGetADCCommand;
import com.qualcomm.hardware.lynx.commands.core.LynxGetADCResponse;
import com.qualcomm.hardware.lynx.commands.core.LynxGetMotorChannelCurrentAlertLevelCommand;
import com.qualcomm.hardware.lynx.commands.core.LynxGetMotorChannelCurrentAlertLevelResponse;
import com.qualcomm.hardware.lynx.commands.core.LynxGetMotorChannelEnableResponse;
import com.qualcomm.hardware.lynx.commands.core.LynxGetMotorChannelModeCommand;
import com.qualcomm.hardware.lynx.commands.core.LynxGetMotorChannelModeResponse;
import com.qualcomm.hardware.lynx.commands.core.LynxGetMotorPIDControlLoopCoefficientsCommand;
import com.qualcomm.hardware.lynx.commands.core.LynxGetMotorPIDControlLoopCoefficientsResponse;
import com.qualcomm.hardware.lynx.commands.core.LynxGetMotorPIDFControlLoopCoefficientsResponse;
import com.qualcomm.hardware.lynx.commands.core.LynxSetMotorChannelCurrentAlertLevelCommand;
import com.qualcomm.hardware.lynx.commands.core.LynxSetMotorConstantPowerCommand;
import com.qualcomm.hardware.lynx.commands.core.LynxSetMotorPIDControlLoopCoefficientsCommand;
import com.qualcomm.hardware.lynx.commands.core.LynxSetMotorPIDFControlLoopCoefficientsCommand;
import com.qualcomm.hardware.lynx.commands.core.LynxSetMotorTargetPositionCommand;
import com.qualcomm.hardware.lynx.commands.core.LynxSetMotorTargetVelocityCommand;
import com.qualcomm.robotcore.hardware.DcMotor;
import com.qualcomm.robotcore.hardware.DcMotorController;
import com.qualcomm.robotcore.hardware.DcMotorEx;
import com.qualcomm.robotcore.hardware.DcMotorImplEx;
import com.qualcomm.robotcore.hardware.MotorControlAlgorithm;
import com.qualcomm.robotcore.hardware.PIDCoefficients;
import com.qualcomm.robotcore.hardware.PIDFCoefficients;
import com.qualcomm.robotcore.hardware.configuration.typecontainers.MotorConfigurationType;
import com.qualcomm.robotcore.util.LastKnown;
import com.qualcomm.robotcore.util.Range;

import org.firstinspires.ftc.robotcore.external.navigation.AngleUnit;
import org.firstinspires.ftc.robotcore.external.navigation.CurrentUnit;
import org.firstinspires.ftc.robotcore.external.navigation.Rotation;
import org.firstinspires.ftc.robotcore.external.navigation.UnnormalizedAngleUnit;

import java.util.concurrent.CompletableFuture;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.TimeoutException;

public class PhotonDcMotor extends DcMotorImplEx {
    private final int port;
    private PhotonHAL hal;
    private final LastKnown<Boolean> enabled;
    private final LastKnown<Integer> setVelocity;
    private final LastKnown<Double> velocity;
    private final LastKnown<PIDCoefficients> pidCoefficients;
    private final LastKnown<PIDFCoefficients> pidfCoefficients;
    private final LastKnown<Double> currentLimit;
    private final LastKnown<ZeroPowerBehavior> zeroPowerBehavior;

    protected Direction direction = Direction.FORWARD;

    protected int positionTolerance = 5;

    protected PhotonDcMotor(PhotonLynxDcMotorController controller, int port, PhotonHAL hal){
        super(controller, port);
        this.port = port;
        this.hal = hal;

        enabled = new LastKnown<>();
        enabled.invalidate();

        setVelocity = new LastKnown<>();
        setVelocity.invalidate();

        velocity = new LastKnown<>();
        velocity.invalidate();

        pidCoefficients = new LastKnown<>();
        pidCoefficients.invalidate();

        pidfCoefficients = new LastKnown<>();
        pidfCoefficients.invalidate();

        currentLimit = new LastKnown<>();
        currentLimit.invalidate();

        zeroPowerBehavior = new LastKnown<>();
        zeroPowerBehavior.invalidate();
    }

    public CompletableFuture<Boolean> isMotorEnabledAsync() {
        CompletableFuture<Boolean> future = new CompletableFuture<>();

        if(enabled.isValid()){
            future = new CompletableFuture<>();
            future.complete(enabled.getValue());
        }else{
            PhotonLynxGetMotorChannelEnableCommand<LynxGetMotorChannelEnableResponse> command = new PhotonLynxGetMotorChannelEnableCommand<>(hal.getLynxModule(), port);
            try {
                future = command.getResponse().thenApply(response -> ((LynxGetMotorChannelEnableResponse)response).isEnabled());
            } catch (LynxNackException e) {
                e.printStackTrace();
            }
            hal.write(command);
        }
        return future;
    }

    public CompletableFuture<PIDCoefficients> getPIDCoefficientsAsync(RunMode mode) {
        CompletableFuture<PIDCoefficients> future = new CompletableFuture<>();
        if(pidCoefficients.isValid()){
            future.complete(pidCoefficients.getValue());
        }else{
            PhotonLynxGetMotorPIDControlLoopCoefficientsCommand command = new PhotonLynxGetMotorPIDControlLoopCoefficientsCommand(hal.getLynxModule(), port, mode);
            hal.write(command);
            try {
                future = command.getResponse().thenApply(message -> {
                    LynxGetMotorPIDControlLoopCoefficientsResponse response = (LynxGetMotorPIDControlLoopCoefficientsResponse) message;
                    pidCoefficients.invalidate();

                    PIDCoefficients coefficients =  new PIDCoefficients(
                            LynxSetMotorPIDControlLoopCoefficientsCommand.externalCoefficientFromInternal(response.getP()),
                            LynxSetMotorPIDControlLoopCoefficientsCommand.externalCoefficientFromInternal(response.getI()),
                            LynxSetMotorPIDControlLoopCoefficientsCommand.externalCoefficientFromInternal(response.getD())
                    );

                    pidCoefficients.updateValue(coefficients);
                    return coefficients;
                });

                return future;
            } catch (LynxNackException e) {
                e.printStackTrace();
            }
        }
        return future;
    }

    public CompletableFuture<Double> getCurrentAsync(CurrentUnit unit) {
        PhotonLynxGetADCCommand command = new PhotonLynxGetADCCommand(hal.getLynxModule(), LynxGetADCCommand.Channel.motorCurrent(port), LynxGetADCCommand.Mode.ENGINEERING);
        try {
            hal.write(command);
            return command.getResponse().thenApply(result -> {
                LynxGetADCResponse response = (LynxGetADCResponse) result;
                return unit.convert(response.getValue(), CurrentUnit.MILLIAMPS);
            });
        } catch (LynxNackException e) {
            return null;
        }
    }

    public double getCorrectedCurrent(CurrentUnit unit) {
        return getCurrent(unit) * Math.abs(getPower());
    }

    public CompletableFuture<Double> getCorrectedCurrentAsync(CurrentUnit unit) {
        PhotonLynxGetADCCommand command = new PhotonLynxGetADCCommand(hal.getLynxModule(), LynxGetADCCommand.Channel.motorCurrent(port), LynxGetADCCommand.Mode.ENGINEERING);
        try {
            hal.write(command);
            return command.getResponse().thenApply(result -> {
                LynxGetADCResponse response = (LynxGetADCResponse) result;
                return unit.convert(response.getValue() * Math.abs(getPower()), CurrentUnit.MILLIAMPS);
            });
        } catch (LynxNackException e) {
            return null;
        }
    }

    public CompletableFuture<Double> getCurrentAlertAsync(CurrentUnit unit) {
        if(currentLimit.isValid()){
            CompletableFuture<Double> future = new CompletableFuture<>();
            future.complete(unit.convert(currentLimit.getValue(), CurrentUnit.MILLIAMPS));
            return future;
        }
        PhotonLynxGetMotorChannelCurrentAlertLevelCommand command = new PhotonLynxGetMotorChannelCurrentAlertLevelCommand(hal.getLynxModule(), port);
        try {
            hal.write(command);
            return command.getResponse().thenApply(result -> {
                LynxGetMotorChannelCurrentAlertLevelResponse response = (LynxGetMotorChannelCurrentAlertLevelResponse) result;
                currentLimit.setValue((double) response.getCurrentLimit());
                return unit.convert(response.getCurrentLimit(), CurrentUnit.MILLIAMPS);
            });
        } catch (LynxNackException e) {
            e.printStackTrace();
        }

        CompletableFuture<Double> future = new CompletableFuture<>();
        future.complete(unit.convert(currentLimit.getRawValue(), CurrentUnit.MILLIAMPS));
        return future;
    }

    public CompletableFuture<ZeroPowerBehavior> getZeroPowerBehaviorAsync() {
        CompletableFuture<ZeroPowerBehavior> future = new CompletableFuture<>();
        if(!zeroPowerBehavior.isValid()){
            PhotonLynxGetMotorChannelModeCommand command = new PhotonLynxGetMotorChannelModeCommand(hal.getLynxModule(), port);
            try {
                hal.write(command);
                return command.getResponse().thenApply(result -> {
                    LynxGetMotorChannelModeResponse response = (LynxGetMotorChannelModeResponse) result;
                    zeroPowerBehavior.updateValue(response.getZeroPowerBehavior());
                    return response.getZeroPowerBehavior();
                });
            } catch (LynxNackException e) {
                e.printStackTrace();
                future.complete(ZeroPowerBehavior.UNKNOWN);
            }
        }else{
            future.complete(zeroPowerBehavior.getRawValue());
        }
        return future;
    }

    public CompletableFuture<RunMode> getModeAsync(){
        PhotonLynxGetMotorChannelModeCommand command = new PhotonLynxGetMotorChannelModeCommand(hal.getLynxModule(), port);
        try {
            hal.write(command);
            return command.getResponse().thenApply(result -> {
                LynxGetMotorChannelModeResponse response = (LynxGetMotorChannelModeResponse) result;
                return response.getMode();
            });
        } catch (LynxNackException e) {
            e.printStackTrace();
            return null;
        }
    }
}

File: PhotonCore/src/main/java/com/outoftheboxrobotics/photoncore/HAL/Motors/PhotonDcMotor.kt
package com.outoftheboxrobotics.photoncore.HAL.Motors

import com.outoftheboxrobotics.photoncore.HAL.HAL
import com.outoftheboxrobotics.photoncore.HAL.Motors.Commands.PhotonLynxGetMotorChannelEnableCommand
import com.outoftheboxrobotics.photoncore.HAL.Motors.Commands.PhotonLynxGetMotorPIDControlLoopCoefficientsCommand
import com.outoftheboxrobotics.photoncore.HAL.Motors.Commands.PhotonLynxSetMotorChannelEnableCommand
import com.outoftheboxrobotics.photoncore.HAL.Motors.Commands.PhotonLynxSetMotorPIDFControlLoopCoefficientsCommand
import com.outoftheboxrobotics.photoncore.HAL.Motors.Commands.PhotonLynxSetMotorTargetVelocityCommand
import com.qualcomm.hardware.lynx.LynxNackException
import com.qualcomm.hardware.lynx.commands.LynxMessage
import com.qualcomm.hardware.lynx.commands.core.*
import com.qualcomm.robotcore.hardware.*
import com.qualcomm.robotcore.hardware.DcMotor.RunMode
import com.qualcomm.robotcore.hardware.DcMotor.ZeroPowerBehavior
import com.qualcomm.robotcore.hardware.HardwareDevice.Manufacturer
import com.qualcomm.robotcore.hardware.configuration.typecontainers.MotorConfigurationType
import com.qualcomm.robotcore.util.LastKnown
import com.qualcomm.robotcore.util.Range
import org.firstinspires.ftc.robotcore.external.navigation.AngleUnit
import org.firstinspires.ftc.robotcore.external.navigation.CurrentUnit
import org.firstinspires.ftc.robotcore.external.navigation.UnnormalizedAngleUnit
import java.util.concurrent.CompletableFuture
import java.util.concurrent.ExecutionException
import java.util.concurrent.TimeUnit
import java.util.concurrent.TimeoutException
import kotlin.math.roundToInt

class PhotonDcMotor protected constructor(private val hal: HAL, private val port: Int) : DcMotorEx {
    private val enabled: LastKnown<Boolean>
    private val setVelocity: LastKnown<Int>
    private val velocity: LastKnown<Double>
    private val pidCoefficients: LastKnown<PIDCoefficients>
    private val pidfCoefficients: LastKnown<PIDFCoefficients>
    protected var direction = DcMotorSimple.Direction.FORWARD

    init {
        enabled = LastKnown()
        enabled.invalidate()
        setVelocity = LastKnown()
        setVelocity.invalidate()
        velocity = LastKnown()
        velocity.invalidate()
        pidCoefficients = LastKnown()
        pidCoefficients.invalidate()
        pidfCoefficients = LastKnown()
        pidfCoefficients.invalidate()
    }

    private fun adjustAngularRate(angularRate: Double): Double {
        var angularRate = angularRate
        if (direction == DcMotorSimple.Direction.REVERSE) angularRate = -angularRate
        return angularRate
    }

    override fun setMotorEnable() {
        if (enabled.updateValue(true)) {
            val command = PhotonLynxSetMotorChannelEnableCommand(
                hal.lynxModule,
                port, true
            )
            hal.write(command)
        }
    }

    override fun setMotorDisable() {
        if (enabled.updateValue(false)) {
            val command = PhotonLynxSetMotorChannelEnableCommand(
                hal.lynxModule,
                port, false
            )
            hal.write(command)
        }
    }

    override fun isMotorEnabled(): Boolean {
        if (enabled.isValid) {
            return enabled.value
        } else {
            val command: PhotonLynxGetMotorChannelEnableCommand<LynxGetMotorChannelEnableResponse> =
                PhotonLynxGetMotorChannelEnableCommand(
                    hal.lynxModule,
                    port
                )
            try {
                val message: LynxMessage = command.getResponse().get(1, TimeUnit.SECONDS)
                enabled.updateValue((message as LynxGetMotorChannelEnableResponse).isEnabled)
                return message.isEnabled
            } catch (e: ExecutionException) {
                e.printStackTrace()
            } catch (e: InterruptedException) {
                e.printStackTrace()
            } catch (e: TimeoutException) {
                e.printStackTrace()
            } catch (e: LynxNackException) {
                e.printStackTrace()
            }
            hal.write(command)
        }
        return enabled.nonTimedValue
    }

    val isMotorEnabledAsync: CompletableFuture<Boolean?>
        get() {
            var future = CompletableFuture<Boolean?>()
            if (enabled.isValid) {
                future = CompletableFuture()
                future.complete(enabled.value)
            } else {
                val command: PhotonLynxGetMotorChannelEnableCommand<LynxGetMotorChannelEnableResponse> =
                    PhotonLynxGetMotorChannelEnableCommand(
                        hal.lynxModule,
                        port
                    )
                try {
                    future = command.getResponse()
                        .thenApply { response -> (response as LynxGetMotorChannelEnableResponse).isEnabled }
                } catch (e: LynxNackException) {
                    e.printStackTrace()
                }
                hal.write(command)
            }
            return future
        }

    override fun setVelocity(angularRate: Double) {
        var angularRate = angularRate
        angularRate = adjustAngularRate(angularRate)
        when (mode) {
            RunMode.RUN_USING_ENCODER, RunMode.RUN_TO_POSITION -> {}
            else -> mode = RunMode.RUN_USING_ENCODER
        }
        val iTicksPerSecond = Range.clip(angularRate.roundToInt(), LynxSetMotorTargetVelocityCommand.apiVelocityFirst, LynxSetMotorTargetVelocityCommand.apiVelocityLast)
        try {
            if (setVelocity.updateValue(iTicksPerSecond)) {
                val command = PhotonLynxSetMotorTargetVelocityCommand(
                    hal.lynxModule,
                    port, iTicksPerSecond
                )
                hal.write(command)
                setMotorEnable()
            }
        } catch (e: RuntimeException) {
            e.printStackTrace()
        }
    }

    /**
     * WARNING: THIS METHOD IS NOT IMPLEMENTED CORRECTLY
     *
     * THIS METHOD WILL ASSUME A GOBILDA 5202 MOTOR
     * USE WITH CAUTION
     */
    override fun setVelocity(angularRate: Double, unit: AngleUnit) {
        val degreesPerSecond = UnnormalizedAngleUnit.DEGREES.fromUnit(unit.unnormalized, angularRate)
        val revolutionsPerSecond = degreesPerSecond / 360.0
        val ticksPerSecond = 28 * revolutionsPerSecond
        setVelocity(ticksPerSecond)
    }

    override fun getVelocity(): Double {
        return if (velocity.isValid) {
            velocity.value
        } else {
            val vel = hal.bulkData!!.getVelocity(port).toDouble()
            velocity.updateValue(vel)
            vel
        }
    }

    override fun getVelocity(unit: AngleUnit): Double {
        val ticksPerSecond = getVelocity()
        val revsPerSecond = ticksPerSecond / 28.0
        return unit.unnormalized.fromDegrees(revsPerSecond * 360.0)
    }

    override fun setPIDCoefficients(mode: RunMode, pidCoefficients: PIDCoefficients) {
        setPIDFCoefficients(mode, PIDFCoefficients(pidCoefficients))
    }

    @Throws(UnsupportedOperationException::class)
    override fun setPIDFCoefficients(mode: RunMode, pidfCoefficients: PIDFCoefficients) {
        if (!mode.isPIDMode) {
            throw RuntimeException("RunMode " + mode.name + " is not a PID mode")
        }
        val command = PhotonLynxSetMotorPIDFControlLoopCoefficientsCommand(
            hal.lynxModule,
            port,
            mode,
            LynxSetMotorPIDControlLoopCoefficientsCommand.internalCoefficientFromExternal(pidfCoefficients.p),
            LynxSetMotorPIDControlLoopCoefficientsCommand.internalCoefficientFromExternal(pidfCoefficients.i),
            LynxSetMotorPIDControlLoopCoefficientsCommand.internalCoefficientFromExternal(pidfCoefficients.d),
            LynxSetMotorPIDControlLoopCoefficientsCommand.internalCoefficientFromExternal(pidfCoefficients.f),
            LynxSetMotorPIDFControlLoopCoefficientsCommand.InternalMotorControlAlgorithm.fromExternal(pidfCoefficients.algorithm)
        )
        hal.write(command)
    }

    override fun setVelocityPIDFCoefficients(p: Double, i: Double, d: Double, f: Double) {
        setPIDFCoefficients(RunMode.RUN_USING_ENCODER, PIDFCoefficients(p, i, d, f, MotorControlAlgorithm.PIDF))
    }

    override fun setPositionPIDFCoefficients(p: Double) {
        setPIDFCoefficients(RunMode.RUN_TO_POSITION, PIDFCoefficients(p, 0.0, 0.0, 0.0, MotorControlAlgorithm.PIDF))
    }

    override fun getPIDCoefficients(mode: RunMode): PIDCoefficients {
        if (pidCoefficients.isValid) {
            return pidCoefficients.value
        } else {
            val command = PhotonLynxGetMotorPIDControlLoopCoefficientsCommand(
                hal.lynxModule,
                port, mode
            )
            hal.write(command)
            try {
                val message: LynxMessage = command.getResponse().get(1, TimeUnit.SECONDS)
                val response = message as LynxGetMotorPIDControlLoopCoefficientsResponse
                pidCoefficients.invalidate()
                val coefficients = PIDCoefficients(
                    LynxSetMotorPIDControlLoopCoefficientsCommand.externalCoefficientFromInternal(response.p),
                    LynxSetMotorPIDControlLoopCoefficientsCommand.externalCoefficientFromInternal(response.i),
                    LynxSetMotorPIDControlLoopCoefficientsCommand.externalCoefficientFromInternal(response.d)
                )
                pidCoefficients.updateValue(coefficients)
                return coefficients
            } catch (e: ExecutionException) {
                e.printStackTrace()
            } catch (e: InterruptedException) {
                e.printStackTrace()
            } catch (e: TimeoutException) {
                e.printStackTrace()
            } catch (e: LynxNackException) {
                e.printStackTrace()
            }
        }
        return pidCoefficients.rawValue
    }

    override fun getPIDFCoefficients(mode: RunMode?): PIDFCoefficients {
        TODO("Not yet implemented")
    }

    override fun setTargetPositionTolerance(tolerance: Int) {
        TODO("Not yet implemented")
    }

    override fun getTargetPositionTolerance(): Int {
        TODO("Not yet implemented")
    }

    override fun getCurrent(unit: CurrentUnit?): Double {
        TODO("Not yet implemented")
    }

    override fun getCurrentAlert(unit: CurrentUnit?): Double {
        TODO("Not yet implemented")
    }

    override fun setCurrentAlert(current: Double, unit: CurrentUnit?) {
        TODO("Not yet implemented")
    }

    override fun isOverCurrent(): Boolean {
        TODO("Not yet implemented")
    }

    override fun getManufacturer(): Manufacturer {
        TODO("Not yet implemented")
    }

    override fun getDeviceName(): String {
        TODO("Not yet implemented")
    }

    override fun getConnectionInfo(): String {
        TODO("Not yet implemented")
    }

    override fun getVersion(): Int {
        TODO("Not yet implemented")
    }

    override fun resetDeviceConfigurationForOpMode() {
        TODO("Not yet implemented")
    }

    override fun close() {
        TODO("Not yet implemented")
    }

    override fun setDirection(direction: DcMotorSimple.Direction?) {
        TODO("Not yet implemented")
    }

    override fun getDirection(): DcMotorSimple.Direction {
        TODO("Not yet implemented")
    }

    override fun setPower(power: Double) {
        TODO("Not yet implemented")
    }

    override fun getPower(): Double {
        TODO("Not yet implemented")
    }

    override fun getMotorType(): MotorConfigurationType {
        TODO("Not yet implemented")
    }

    override fun setMotorType(motorType: MotorConfigurationType?) {
        TODO("Not yet implemented")
    }

    override fun getController(): DcMotorController {
        TODO("Not yet implemented")
    }

    override fun getPortNumber(): Int {
        TODO("Not yet implemented")
    }

    override fun setZeroPowerBehavior(zeroPowerBehavior: ZeroPowerBehavior?) {
        TODO("Not yet implemented")
    }

    override fun getZeroPowerBehavior(): ZeroPowerBehavior {
        TODO("Not yet implemented")
    }

    override fun setPowerFloat() {
        TODO("Not yet implemented")
    }

    override fun getPowerFloat(): Boolean {
        TODO("Not yet implemented")
    }

    override fun setTargetPosition(position: Int) {
        TODO("Not yet implemented")
    }

    override fun getTargetPosition(): Int {
        TODO("Not yet implemented")
    }

    override fun isBusy(): Boolean {
        TODO("Not yet implemented")
    }

    override fun getCurrentPosition(): Int {
        TODO("Not yet implemented")
    }

    override fun setMode(mode: RunMode?) {
        TODO("Not yet implemented")
    }

    override fun getMode(): RunMode {
        TODO("Not yet implemented")
    }
}